File: csrc/vstack.cpp
#include <stdio.h>
#include "tokeniterator.h"
#include "quotestream.h"
#include "cmdquote.h"
#include "vexception.h"
#include "vstack.h"
VStack::VStack():_now(new (collect) Node(0)),_first(_now) {
}

Node* VStack::now() {
    return _now;
}

Node* VStack::now(Node* n) {
    return _now = n;
}

Token* VStack::push(Token* t) {
    Node_ n = new (collect) Node(t);
    n->link = _now;
    _now = n;
    return _now->data;
}

Token* VStack::pop() {
    if (!_now || !_now->data)
        throw VException("err:stack_empty", new (collect) Term(TInt, (long)0), "Empty Stack.");
    Token_ t = _now->data;
    _now = _now->link;
    return t;
}

bool VStack::empty() {
    return _now->link == 0;
}

void VStack::clear() {
    _now = _first;
}

Token* VStack::peek() {
    return _now->data;
}

Node* VStack::getList() {
    Node_ current = _now;
    Node_ result = 0;
    Node_ t = 0;
    while (current && current->data) {
        t = new (collect) Node(current->data);
        t->link = result;
        result = t;
        current = current->link;
    }
    return result;
}

Quote* VStack::quote() {
    Node_ s = getList();
    QuoteStream_ qs = new (collect) QuoteStream();
    while(s) {
        qs->add(s->data);
        s = s->link;
    }
    return new (collect) CmdQuote(qs);
}

void VStack::dequote(Quote* q) {
    Node_ current = _now;
    TokenIterator_ it = q->tokens()->iterator();

    _now = new (collect) Node(0);
    _first = _now;

    while(it->hasNext())
        push(it->next());
}

void VStack::dump() {
    Node_ s = getList();
    printf("(");
    bool first = true;
    while(s) {
        printf("%s%s", first ? "" : " ",s->data->value());
        first = false;
        s = s->link;
    }
    printf(")\n");
}

File: v/VStack.java
package v;
import java.util.*;

class Node<E> {
    E data;
    Node<E> link;
    public Node(E e) {
        data = e;
    }
}

public class VStack {
    Node<Term> now = null;
    Node<Term> first = null;
    public VStack() {
        now = new Node<Term>(null);
        first = now;
    }

    public Node<Term> now() {
        return now;
    }

    public Node<Term> now(Node<Term> n) {
        now = n;
        return n;
    }

    public Term push(Term t) {
        Node<Term> n = new Node<Term>(t);
        n.link = now;
        now = n;
        return now.data;
    }

    public Term pop() {
        if (now == null || now.data == null)
            throw new VException("err:stack_empty","Empty Stack.");
        Term t = now.data;
        now = now.link;
        return t;
    }

    public boolean empty() {
        return now.link == null;
    }

    public void clear() {
        now = first;
    }

    public Term peek() {
        return now.data;
    }

    private List<Term> getlist() {
        LinkedList<Term> l = new LinkedList<Term>();
        Node<Term> current = now;
        while(current != null && current.link != null) {
            l.addFirst(current.data);
            current = current.link;
        }
        return l;
    }

    public Quote quote() {
        QuoteStream qs = new QuoteStream();
        List<Term> l = getlist();
        for(Term t: l)
            qs.add(t);
        return new CmdQuote(qs);
    }
  
    public void dequote(Quote q) {
        Node<Term> current = now;
        Iterator<Term> it = (Iterator<Term>)q.tokens().iterator();

        now = new Node<Term>(null);
        first = now;

        while(it.hasNext()) {
            push(it.next());
        }
    }

    public void dump() {
        List<Term> l = getlist();
        V.out("(");
        boolean first = true;
        for(Term t: l) {
            if (first) first = false;
            else V.out(" ");
            switch (t.type) {
                case TString:
                    V.out("'" + t.svalue() + "'");
                    break;
                case TChar:
                    V.out("~" + t.svalue());
                    break;
                default:
                    V.out(t.value());
            }
        }
        V.outln(")");
    }
}
